Add support on omitempty for all types

Hello,

I expected this behaviour which match `encoding/json` package.

Example: https://play.golang.org/p/o6IiMmNMK9

I wrote below the test to match the behaviour above but it fails.

The unit test:

```go

type E5 struct {
	F01 int     `schema:"f01,omitempty"`
	F02 string  `schema:"f02,omitempty"`
	F03 *string `schema:"f03,omitempty"`
	F04 *int8   `schema:"f04,omitempty"`
	F05 float64 `schema:"f05,omitempty"`
}

func TestEncoderWithOmitempty(t *testing.T) {
	vals := map[string][]string{}

	s := E5{}

	encoder := NewEncoder()
	encoder.Encode(&s, vals)

	valNotExists(t, "f01", vals)
	valNotExists(t, "f02", vals)
	valNotExists(t, "f03", vals)
	valNotExists(t, "f04", vals)
	valNotExists(t, "f05", vals)
}
```

Here is the trace:

```
Running tool: /usr/local/bin/go test -timeout 30s -tags  -run ^TestEncoderWithOmitempty$

--- FAIL: TestEncoderWithOmitempty (0.00s)
	/Users/aetcheverry/Projects/Go/src/github.com/gorilla/schema/encoder_test.go:273: Key not ommited. Expected: empty; got: 0.
	/Users/aetcheverry/Projects/Go/src/github.com/gorilla/schema/encoder_test.go:273: Key not ommited. Expected: empty; got: null.
	/Users/aetcheverry/Projects/Go/src/github.com/gorilla/schema/encoder_test.go:273: Key not ommited. Expected: empty; got: null.
	/Users/aetcheverry/Projects/Go/src/github.com/gorilla/schema/encoder_test.go:273: Key not ommited. Expected: empty; got: 0.000000.
FAIL
exit status 1
FAIL	github.com/gorilla/schema	0.005s
Error: Tests failed.
```

Thanks for your help. 😃

--- a/encoder.go
+++ b/encoder.go
@@ -67,7 +67,7 @@ func (e *Encoder) encode(v reflect.Value, dst map[string][]string) error {
 		// Encode non-slice types and custom implementations immediately.
 		if encFunc != nil {
 			value := encFunc(v.Field(i))
-			if value == "" && opts.Contains("omitempty") {
+			if opts.Contains("omitempty") {
 				continue
 			}
 

--- a/encoder.go
+++ b/encoder.go
@@ -67,7 +67,7 @@ func (e *Encoder) encode(v reflect.Value, dst map[string][]string) error {
 		// Encode non-slice types and custom implementations immediately.
 		if encFunc != nil {
 			value := encFunc(v.Field(i))
-			if opts.Contains("omitempty") {
+			if v.Field(i).Interface() == reflect.Zero(reflect.TypeOf(v.Field(i).Interface())).Interface() && opts.Contains("omitempty") {
 				continue
 			}
 

--- a/encoder.go
+++ b/encoder.go
@@ -67,7 +67,7 @@ func (e *Encoder) encode(v reflect.Value, dst map[string][]string) error {
 		// Encode non-slice types and custom implementations immediately.
 		if encFunc != nil {
 			value := encFunc(v.Field(i))
-			if v.Field(i).Interface() == reflect.Zero(reflect.TypeOf(v.Field(i).Interface())).Interface() && opts.Contains("omitempty") {
+			if opts.Contains("omitempty") && v.Field(i).Interface() == reflect.Zero(reflect.TypeOf(v.Field(i).Interface())).Interface() {
 				continue
 			}
 

--- a/encoder_test.go
+++ b/encoder_test.go
@@ -296,19 +296,32 @@ type E5 struct {
 	F03 *string `schema:"f03,omitempty"`
 	F04 *int8   `schema:"f04,omitempty"`
 	F05 float64 `schema:"f05,omitempty"`
+	F06 E5F06   `schema:"f06,omitempty"`
+	F07 E5F06   `schema:"f07,omitempty"`
+}
+
+type E5F06 struct {
+	F0601 string `schema:"f0601,omitempty"`
 }
 
 func TestEncoderWithOmitempty(t *testing.T) {
 	vals := map[string][]string{}
 
-	s := E5{}
+	s := E5{
+		F02: "test",
+		F07: E5F06{
+			F0601: "test",
+		},
+	}
 
 	encoder := NewEncoder()
 	encoder.Encode(&s, vals)
 
 	valNotExists(t, "f01", vals)
-	valNotExists(t, "f02", vals)
+	valExists(t, "f02", "test", vals)
 	valNotExists(t, "f03", vals)
 	valNotExists(t, "f04", vals)
 	valNotExists(t, "f05", vals)
+	valNotExists(t, "f06", vals)
+	valExists(t, "f0601", "test", vals)
 }

--- a/encoder_test.go
+++ b/encoder_test.go
@@ -291,13 +291,15 @@ func TestEncoderSetAliasTag(t *testing.T) {
 }
 
 type E5 struct {
-	F01 int     `schema:"f01,omitempty"`
-	F02 string  `schema:"f02,omitempty"`
-	F03 *string `schema:"f03,omitempty"`
-	F04 *int8   `schema:"f04,omitempty"`
-	F05 float64 `schema:"f05,omitempty"`
-	F06 E5F06   `schema:"f06,omitempty"`
-	F07 E5F06   `schema:"f07,omitempty"`
+	F01 int      `schema:"f01,omitempty"`
+	F02 string   `schema:"f02,omitempty"`
+	F03 *string  `schema:"f03,omitempty"`
+	F04 *int8    `schema:"f04,omitempty"`
+	F05 float64  `schema:"f05,omitempty"`
+	F06 E5F06    `schema:"f06,omitempty"`
+	F07 E5F06    `schema:"f07,omitempty"`
+	F08 []string `schema:"f08,omitempty"`
+	F09 []string `schema:"f09,omitempty"`
 }
 
 type E5F06 struct {
@@ -312,6 +314,7 @@ func TestEncoderWithOmitempty(t *testing.T) {
 		F07: E5F06{
 			F0601: "test",
 		},
+		F09: []string{"test"},
 	}
 
 	encoder := NewEncoder()
@@ -324,4 +327,6 @@ func TestEncoderWithOmitempty(t *testing.T) {
 	valNotExists(t, "f05", vals)
 	valNotExists(t, "f06", vals)
 	valExists(t, "f0601", "test", vals)
+	valNotExists(t, "f08", vals)
+	valsExist(t, "f09", []string{"test"}, vals)
 }

--- a/encoder_test.go
+++ b/encoder_test.go
@@ -289,3 +289,26 @@ func TestEncoderSetAliasTag(t *testing.T) {
 	encoder.Encode(&s, data)
 	valExists(t, "id", "foo", data)
 }
+
+type E5 struct {
+	F01 int     `schema:"f01,omitempty"`
+	F02 string  `schema:"f02,omitempty"`
+	F03 *string `schema:"f03,omitempty"`
+	F04 *int8   `schema:"f04,omitempty"`
+	F05 float64 `schema:"f05,omitempty"`
+}
+
+func TestEncoderWithOmitempty(t *testing.T) {
+	vals := map[string][]string{}
+
+	s := E5{}
+
+	encoder := NewEncoder()
+	encoder.Encode(&s, vals)
+
+	valNotExists(t, "f01", vals)
+	valNotExists(t, "f02", vals)
+	valNotExists(t, "f03", vals)
+	valNotExists(t, "f04", vals)
+	valNotExists(t, "f05", vals)
+}

